[vcpkg docs] new README/Quick Start guide (#12093)
microsoft/vcpkg: README.md

# Vcpkg: Overview

Vcpkg helps you manage C and C++ libraries on Windows, Linux and MacOS.

> This tool and ecosystem are constantly evolving; your involvement is vital to its success!

Reviewer: ```suggestion
This tool and ecosystem are constantly evolving; your involvement is greatly appreciated!
```

This is preexisting but I think this statement sounds like 'if people don't help we're going to give up on it' which isn't true anymore. Up to you to take or leave.

# Table of Contents

- [Vcpkg: Overview](#vcpkg-overview)
- [Table of Contents](#table-of-contents)
- [Getting Started](#getting-started)
  - [Quick Start: Windows](#quick-start-windows)
  - [Quick Start: Unix](#quick-start-unix)
  - [Installing Linux Developer Tools](#installing-linux-developer-tools)
  - [Installing macOS Developer Tools](#installing-macos-developer-tools)
    - [Installing GCC on macOS](#installing-gcc-on-macos)
  - [Using vcpkg with CMake](#using-vcpkg-with-cmake)
    - [Vcpkg as a Submodule](#vcpkg-as-a-submodule)
    - [Visual Studio Code with CMake Tools](#visual-studio-code-with-cmake-tools)
  - [Quick Start: Manifests](#quick-start-manifests)
- [Tab-Completion/Auto-Completion](#tab-completionauto-completion)
- [Examples](#examples)
- [Contributing](#contributing)
- [License](#license)
- [Telemetry](#telemetry)

# Getting Started

First, follow the quick start guide for either
[Windows](#quick-start-windows), or [macOS and Linux](#quick-start-unix),
depending on what you're using.

For more information, see [Installing and Using Packages][getting-vcpkg:using-a-package].

> If a library you need is not present in vcpkg catalog,

Author: How do you mean?

> If a library you need is not present in vcpkg catalog,

Reviewer 2: No, it makes sense, marking as resolved

you can [open an issue on the GitHub repo][getting-vcpkg:open-issue]
where the vcpkg team and community can see it,
and potentially add the port to vcpkg.

After you've gotten vcpkg installed and working,
you may wish to add [tab completion](#tab-completionauto-completion) to your shell.

Finally, if you're interested in the future of vcpkg,
check out the [manifest](#quick-start-manifest) guide!
This is an experimental feature and will likely have bugs,
so try it out and [open all the issues][contributing:submit-issue]!

## Quick Start: Windows

Prerequisites:
- Windows 7 or newer
- [Git][quick-start:git]
- [Visual Studio][quick-start:visual-studio] 2015 Update 3 or greater with the English language pack

First, download and bootstrap vcpkg itself; it can be installed anywhere,
but generally we recommend using vcpkg as a submodule for CMake projects,

> and installing it on a user-local level for Visual Studio projects.

Reviewer 3: Because of path length, space, and unicode issues, we should recommend a path such as `C:\dev\vcpkg` or `C:\src\vcpkg`.

```cmd
> git clone https://github.com/microsoft/vcpkg
> .\vcpkg\bootstrap-vcpkg.bat
```

To install the libraries for your project, run:

```cmd

> > .\vcpkg\vcpkg install [packages to install] # on Windows

Reviewer 3: Since this is the Windows section, the `# on Windows` probably isn't needed.

```

In order to use vcpkg with Visual Studio, including CMake projects in Visual Studio,

> run the following command in an administrator command prompt:

Reviewer 3: Users shouldn't run this in an admin prompt -- we'll pop up an elevation request if needed.

```cmd
> .\vcpkg\vcpkg integrate install
```

> After this, you can now create a New Project (or open an existing one).

Reviewer 3: I think users might be confused between MSBuild and CMake here; we should directly users in all of our main scenarios (VS MSBuild, VS CMake, VSCode CMake, CMake CLI)

Author: I don't understand this sentence

Reviewer 3: "New Project" could mean MSBuild projects or CMake projects

Author: sure, but isn't that okay?

Reviewer 3: In the case of CMake, the next line is not true:
>All installed libraries are immediately ready to be `#include`'d and used
in your project without additional configuration.

If a CMake user stops reading at this point, they will be quite upset that their `#include<>`'s don't magically work without calling `find_package()`

Author: Oh! yeah that's fair

in your project without additional configuration.

In order to use vcpkg with CMake, for non-Visual Studio workflows,
you can use the toolchain file:

```
> cmake -B [build directory] -S . -DCMAKE_TOOLCHAIN_FILE=[path to vcpkg]/scripts/buildsystems/vcpkg.cmake
> cmake --build [build directory]
```

With CMake, you will still need to `find_package` and the like to use the libraries.
Check out the [CMake section](#using-vcpkg-with-cmake) for more information.

For any other tools, including Visual Studio Code,
check out the [integration guide][quick-start:integration].

## Quick Start: Unix

Prerequisites for Linux:
- [Git][quick-start:git]
- [g++][quick-start:linux-gcc] >= 6

Prerequisites for macOS:
- [Apple Developer Tools][quick-start:apple-developer-tools]
- On macOS 10.14 or below, you will also need:
  - [Homebrew][quick-start:homebrew]
  - [g++][quick-start:macos-gcc] >= 6 from Homebrew

First, download and bootstrap vcpkg itself; it can be installed anywhere,
but generally we recommend using vcpkg as a submodule for CMake projects.

```sh
$ git clone https://github.com/microsoft/vcpkg
$ ./vcpkg/bootstrap-vcpkg.sh
```

To install the libraries for your project, run:

```

> > ./vcpkg/vcpkg install [packages to install]

Reviewer 3: `$ ./vcpkg/vcpkg install`

```

In order to use vcpkg with CMake, you can use the toolchain file:

```
> cmake -B [build directory] -S . -DCMAKE_TOOLCHAIN_FILE=[path to vcpkg]/scripts/buildsystems/vcpkg.cmake
> cmake --build [build directory]
```

With CMake, you will still need to `find_package` and the like to use the libraries.
Check out the [CMake section](#using-vcpkg-with-cmake)
for more information on how best to use vcpkg with CMake,
and CMake Tools for VSCode.

For any other tools, check out the [integration guide][quick-start:integration].

## Installing Linux Developer Tools

Across the different distros of Linux, there are different packages you'll
need to install:

- Debian, Ubuntu, popOS, and other Debian-based distributions:

```sh
$ sudo apt-get update

> $ sudo apt-get install build-essential

Reviewer: What about tar, unzip, and curl? (That is, anything bootstrap will need)

```

- CentOS

```sh
$ sudo yum install centos-release-scl
$ sudo yum install devtoolset-7
$ scl enable devtoolset-7 bash
```

For any other distributions, make sure you're installing g++ 6 or above.
If you want to add instructions for your specific distro,
[please open a PR][contributing:submit-pr]!

## Installing macOS Developer Tools

On macOS 10.15, the only thing you should need to do is run the following in your terminal:

```sh
$ xcode-select --install
```

Then follow along with the prompts in the windows that comes up.

On macOS 10.14 and previous, you'll also need to install g++ from homebrew;
follow the instructions in the following section.

> ### Installing GCC on macOS

Reviewer: ```suggestion
### Installing GCC for macOS before 10.15
```

Up to you

This will _only_ be necessary if you're using a macOS version from before 10.15.
Installing homebrew should be very easy; check out <brew.sh> for more information,
but at its simplest, run the following command:

```sh
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
```

Then, in order to grab an up-to-date version of gcc, run the following:

```sh
$ brew install gcc
```

You'll then be able to bootstrap vcpkg along with the [quick start guide](#quick-start-unix)

## Using vcpkg with CMake

If you're using vcpkg with CMake, the following may help!

### Vcpkg as a Submodule

When using vcpkg as a submodule of your project,
you can add the following to your CMakeLists.txt instead of passing `CMAKE_TOOLCHAIN_FILE`

> to the cmake invocation.

Reviewer 4: Needs to be set before the first `project()` call. 
I personally use [find_package(VCPKG)](https://github.com/Neumann-A/CMakeCommonSetup/blob/master/cmake/modules/FindVCPKG.cmake) since for me it seems cleaner and I can add VCPKG related features/variables in an isolated module.

```cmake
set(CMAKE_TOOLCHAIN_FILE ${CMAKE_SOURCE_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake

>   CACHE STRING "Vcpkg toolchain file")

Reviewer 3: I think you should include the `project()` directive as part of this snippet to make it obvious during skimming that this needs to go before `project()`.

Also, this should probably be `${CMAKE_CURRENT_LIST_DIR}` so that `add_subdirectory()` sortof works.

```

This will still allow people to not use vcpkg,
by passing the `CMAKE_TOOLCHAIN_FILE` directly,
but it will make the configure-build step slightly easier.

### Visual Studio Code with CMake Tools

Adding the following to your workspace `settings.json` will make
CMake Tools automatically use vcpkg for libraries:

```json
{
  "cmake.configureSettings": {
    "CMAKE_TOOLCHAIN_FILE": "[vcpkg root]/scripts/buildsystems/vcpkg.cmake"
  }
}
```

## Quick Start: Manifests

So, you want to look at what the future of vcpkg is going to be like!
We really appreciate it 😄. First, though, a warning:

> manifest support in vcpkg is still in beta! Mostly, it should just work,

Reviewer: I don't think we should claim this until your manifests implementation PR lands.

Author: My thought was that they're going to get merged within a few hours of each other, and I don't really want to deal with the whole "one PR depends on another"

Reviewer: That's close enough for me

but there are no guarantees and it's likely you'll hit at least a bug or two
while you're using vcpkg in this mode. Additionally,
we will likely break behavior before stabilizing, so be forewarned.
Please [open issues][contributing:submit-issue] if you hit any bugs!

First, install vcpkg as normal for [Windows](#quick-start-windows) or
[Unix](#quick-start-unix).
You may wish to install vcpkg in a central place,
as the installed directory exists locally,
and it's fine to run multiple vcpkg commands from the same
vcpkg directory at the same time.

Then, we must turn on the `manifests` vcpkg feature flag by adding

> `manifests` to the comma-separated `VCPKG_FEATURE_FLAGS` environment

Reviewer: Stupid ?: Do we want to encourage people to use that environment variable or to use the command line method instead? Environment is more convenient to share across commands but command line is closer to 'pit of success' whenever one must integrate with other features. Up to you.

Author: hmm, fair enough.

although you will need to set `VcpkgManifestEnabled` to `On` for MSBuild.
If you wish to install your dependencies without using either CMake or MSBuild,
you may use a simple `vcpkg install` without arguments.

For more information, check out the [manifest][getting-started:manifest-spec]
specification.

[getting-started:using-a-package]: docs/examples/intalling-and-using-packages.md
[getting-started:git]: https://git-scm.com/downloads
[getting-started:cmake-tools]: https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools
[getting-started:linux-gcc]: #installing-linux-developer-tools
[getting-started:macos-dev-tools]: #installing-macos-developer-tools
[getting-started:macos-brew]: #installing-gcc-on-macos
[getting-started:macos-gcc]: #installing-gcc-on-macos
[getting-started:visual-studio]: https://visualstudio.microsoft.com/
[getting-started:manifest-spec]: docs/specifications/manifests.md

# Tab-Completion/Auto-Completion

`vcpkg` supports auto-completion of commands, package names,
and options in both powershell and bash.
To enable tab-completion in the shell of your choice, run:

```pwsh
> .\vcpkg integrate powershell
```

or

```

> $ ./vcpkg integrate bash

Reviewer 5: ```
```sh
```
